trinity/utils/shutdown.py

```diff
@@ -12,16 +18,23 @@
 
 async def exit_on_signal(service_to_exit: BaseService, endpoint: Endpoint = None) -> None:
     loop = service_to_exit.get_event_loop()
+    async with wait_for_exit(loop):
+        await service_to_exit.cancel()
+        if endpoint is not None:
+            endpoint.stop()
+        service_to_exit._executor.shutdown(wait=True)
```

Comment:
Maybe add another one like:

```py
@asynccontextmanager
async def wait_for_service_exit(service_to_exit: BaseService) -> None:
    loop = service_to_exit.get_event_loop()
    async with wait_for_exit(loop):
        await service_to_exit.cancel()
        yield
        service_to_exit._executor.shutdown(wait=True)
```

`exit_on_signal` might go away or be pushed elsewhere, into `exit_service_with_plugin_endpoint(service, endpoint)` which just stops the endpoint, and `simple_service_exit(service)` which does nothing at all in `wait_for_service_exit`.

Reply:
I went down that way but slightly adjusted naming scheme:

```python
async def exit_with_service_and_endpoint(service_to_exit: BaseService, endpoint: Endpoint) -> None:
    async with exit_signal_with_service(service_to_exit):
        endpoint.stop()


async def exit_with_service(service_to_exit: BaseService) -> None:
    async with exit_signal_with_service(service_to_exit):
        pass


@asynccontextmanager
async def exit_signal_with_service(service_to_exit: BaseService) -> AsyncGenerator[None, None]:
    loop = service_to_exit.get_event_loop()
    async with exit_signal(loop):
        await service_to_exit.cancel()
        yield
        service_to_exit._executor.shutdown(wait=True)


@asynccontextmanager
async def exit_signal(loop: asyncio.AbstractEventLoop) -> AsyncGenerator[None, None]:
    sigint_received = asyncio.Event()
    for sig in [signal.SIGINT, signal.SIGTERM]:
        # TODO also support Windows
        loop.add_signal_handler(sig, sigint_received.set)

    await sigint_received.wait()
    try:
        yield
    finally:
        loop.stop()
```

I think this reads more natural:

```python
async with exit_signal(loop):
    # do something
```

Same with:

```python

asyncio.ensure_future(exit_with_service_and_endpoint(service, endpoint))
```